Zend/zend_execute.c

```diff
@@ -1874,13 +1888,23 @@ static zend_always_inline void zend_fetch_dimension_address_read(zval *result, z
 			}
 		}
 	} else {
-		if (type != BP_VAR_IS && UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
-			zval_undefined_cv(EG(current_execute_data)->opline->op1.var, EG(current_execute_data));
+		if (type != BP_VAR_IS) {
+			if (UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
+				zval_undefined_cv(EG(current_execute_data)->opline->op1.var, EG(current_execute_data));
+			} else if (!zend_fetch_dimension_address_read_valid_container(container, support_strings)) {
+				zend_error(E_WARNING, "Variable of type %s does not accept array offsets", zend_get_type_by_const(Z_TYPE_P(container)));
+			}
 		}
 		if (/*dim_type == IS_CV &&*/ UNEXPECTED(Z_TYPE_P(dim) == IS_UNDEF)) {
 			zval_undefined_cv(EG(current_execute_data)->opline->op2.var, EG(current_execute_data));
 		}
-		ZVAL_NULL(result);
+
+		/* return error if we're continuing looking up.  This is OK since after FETCH_DIM_R there will be a FREE op */
+		if ((EG(current_execute_data)->opline + 1)->opcode == ZEND_FETCH_DIM_R) {
+			ZVAL_ERROR(result);
```

Comment:
It should be possible to resolve that particular problem by checking whether the result var of this op is the op1 of the subsequent FETCH_DIM_R, to avoid this kind of mixup between different, but subsequent, dim fetches. However this is still not a nice solution.